Document the usage of a singularity cache. Encourage users to use the --containall flag with singularity.

rhpvorderman opened the pull request:
This PR documents how to use the singularity cache in a way that every image only gets pulled once.

It also documents why the `--containall` flag should be used with singularity at all times.

Pinging @illusional @TMiguelT and @vsoch as heavy singularity promotors. What do you think of this?

vsoch commented on docs/tutorials/Containers.md:
```diff
@@ -138,14 +138,31 @@ As the configuration will require more knowledge about your execution environmen
 
 As the `Singularity exec` command does not emit a job-id, we must include the `run-in-background` tag within the the provider section in addition to the docker-submit script. As Cromwell watches for the existence of the `rc` file, the `run-in-background` option has the caveat that we require the Singularity container to successfully complete, otherwise the workflow might hang indefinitely.
 
+The `--containall` flag is **important**. By default Singularity will:
```
Interesting suggestion - don't most users _want_ that seamless transition between host and container, especially for writing files? Is there any reason to not suggest pulling the image first and then running exec against the binary?

cjllanwarne replied:
Mounting local filesystem locations without explicitly recording file inputs and outputs in your WDL makes it harder to determine call cache validity and makes the WDL non-portable.

vsoch replied:
Ah, ok - so the context is when using wdl, and yes that makes sense. Could we add that as a clarification so it's not lost to dinosaurs like me that don't read it with this context?

illusional replied:
Yeah I agree with this, I think it’s super important for reproducibility To have this containall flag. I’ve seen random things like python installations break inside a container without it.

I think adding a precursor sentence would be good, something like “to ensure reproducibility and an isolated environment inside the container, ‘containall’ is an **important** function ...”

rhpvorderman replied:
> Is there any reason to not suggest pulling the image first and then running exec against the binary?

Yes. @TMIguelT has documented this in this issue: https://github.com/broadinstitute/cromwell/issues/5063.

Using a cache prevents pulling the same image over and over again. To populate the cache a command needs to be run before the submit command.

TMiguelT replied:
Well actually it probably is ideal to pull the image first (before running cromwell at all), but then it's no longer a download-and-run setup, it requires manual intervention. 

Imagine if you didn't know anything about singularity and WDL at all, and you just wanted to run some WDL you found online, using your pre-configured Cromwell + Singularity. You would have no idea how to work out what containers are used, or even what *containers* are, necessarily, which is why I would like Cromwell to do this for us.

vsoch replied:
If the image is pulled once to a shared filesystem it’s the shared target for all the jobs, like any binary or script. Is the filesystem not shared? What am I missing?

rhpvorderman replied:
@vsoch you are correct. But by using exec we let Singularity do all the work of localizing and storing the images instead of doing it ourselves in bash.

I prefer to let tools work for me, instead if me working for the tools :wink:.

TMiguelT replied:
Yes so if it's pulled once it will work and be shared across all workers. The problem is that pulling manually requires knowledge of singularity, and pulling automatically inside Cromwell requires a file lock to prevent wacky race conditions where all workers are simultaneously writing to the cache.

rhpvorderman replied:
Well I wouldn't desribe this filelock as a problem. I mean, it solves the problem nicely for the singularity use case. Given the design of `flock` and Cromwell I  am quite certain this effectively solves all problems.

vsoch replied:
Ah understood. For clusters not using Cromwell, there has been a little bit of disaster to not pull first (and having nodes each starting to pull a 500GB container at the same time) so I advocate for puling first. One more question - if we always exec with a particular container URI and the latest (or tag) changes, Singularity will detect a difference in the sha and  pull again. The user might not be aware that the container has changed. Pulling first and interacting with the binary solves this issue.

TMiguelT replied:
Encouraging users to use the container hash? 😬

vsoch replied:
I think that kind of usage is hard for most, you have to look it up or know how to inspect Docker. If you could figure out the percentage of users that pull with a hash I’d suspect it to be very small. A pull before running a job to produce a binary that you confidently use (and reuse) is still my preference here, although I suspect that’s not the Cromwell way :)

rhpvorderman replied:
@vsoch. This does pull first! It simply uses the exec command to do so, because that way singularity will do the pull and use the cache. It uses a different cache for `exec` than it `uses` for pull (the layers are the same, but with exec it also creates a build image in the cache.)

rhpvorderman replied:
Only AFTER the exec statement is the job submitted.

rhpvorderman replied:
Also @TMiguelT and @vsoch doesn't cromwell use the container hash by default? And stores this in the database?

Also if you are a decent person and use biocontainers instead of rolling your own, you should not get any changing tags, as the build system for biocontainers does not allow this.

TMiguelT replied:
I don't know how the Cromwell internals work. Certainly this would be ideal.

I also didn't realise biocontainers don't override tags, but still many images from external sources (`miniconda:latest`, or `ubuntu:latest`, for example) risk changing occasionally.

rhpvorderman replied:
> I don't know how the Cromwell internals work. Certainly this would be ideal.

There is quite some code in cromwell to handle docker specifics. Including hash lookup.

> I also didn't realise biocontainers don't override tags, but still many images from external sources (miniconda:latest, or ubuntu:latest, for example) risk changing occasionally.

That is true. In biowdl tasks we only use biocontainers for this reason. A tag is more readable than a hash. For some corner case stuff we use a debian container, but that one is fixed at a specific sha256 hash. 

But anyway, we are digressing :wink:. This does not matter much for this specific PR, I think.

TMiguelT replied:
It matters only in the sense that this config doesn't guarantee reproducibility, but it's probably not its job to do so.

rhpvorderman replied:
Agreed. But indeed, no config will guarantee reproducibility of the user chooses the wrong containers :confused:.

rhpvorderman replied:
As regards to this, should we also give a mention to the prepull-singularity script from BioWDL? I have written the config so that it is not needed, but maybe users want it is an option? https://github.com/biowdl/prepull-singularity

vsoch commented on docs/tutorials/Containers.md:
```diff
@@ -178,27 +195,48 @@ To run Singularity on a job scheduler, the singularity command needs to be passe
-- We'll use the docker name (potentially including the docker hash) to generate a compliant filename that can be looked up later.
-- If the container exists, there is no need to rebuild the container.
+- Make sure Singularity is loaded (and in our path), for example you can 
+  call `module load Singularity/3.0.1` if module is installed on your system.
```
This version is fairly old - any reason to use it for the example?

rhpvorderman replied:
I did not put it there. This was already there and I did not change it. It may be a good thing to remove things that reference version as these get outdated eventually.

To be frank. I wonder why the `module load` singularity is there in the first place as you can also use the path to the singularity executable. For example `/clusterfs/shared-software/singularity/3.5.3/bin/singularity`.

vsoch replied:
The module load sometimes also sets SINGULARITY_CACHEDIR or other envars.

TMiguelT replied:
While we're here, we might as well bump the singularity version to something newer but still stable.

illusional replied:
My thought when we wrote it was, people are using slurm which means there's a good chance they're using or are familiar with modules.

rhpvorderman replied:
`module load Singularity` without a version should default to the latest version. So the version statement can be dropped altogether.

vsoch commented on docs/tutorials/Containers.md:
```diff
@@ -178,27 +195,48 @@ To run Singularity on a job scheduler, the singularity command needs to be passe
+  pulling the image before the job is submitted. The flock and pull command 
+  needs to be placed *before* the submit command so all pull commands are 
+  executed on the same node. This is necessary for the filelock to work.
+- As mentioned above the `--containall` flag is **important** for 
```
I'm not sure I totally agree with this. If you can't write files, that's a problem. containall becomes important when you have issues with path conflicts, etc.

rhpvorderman replied:
Containers implicitly loading the user's environment and home directory which may affect settings in certain tools is  a big problem. For reproducibility I only want settings that are explicitly given on the command line or otherwise via the WDL task to take effect.
Otherwise different users may have different results, meaning the workflow is not reproducable. In scientific workloads that is *completely unacceptable*. The use `--containall` recommendation can not be strong enough.
If you need functionality that is disabled by `--containall` you can always explicitly request this functionality. This is better in that you make your requirements explicit! Implicit requirements make it very hard to reproduce workflows.

vsoch replied:
Cant argue with that. :) Why not mention —cleanenv too then?

rhpvorderman replied:
`--containall` implies `--cleanenv`. As well as `--contain` (no HOME mounting), `--ipc` (run in a new IPC namespace) and `--pid` (run in a new PID namespace). 

Technically using `--cleanenv --contain` also works. Except that some python tools (as @illusional mentioned) crash. This is probably (I am not certain) due to these python tools using the multiprocessing module which requires a separate IPC namespace. So `--ipc` and `--pid` should also be added to be sure.

Using `--containall` might also be translated as "Do not implicitly inherited anything from the host except the bare necessities. If I need more I will tell you so explicitly". I think that is good default behavior for containers to observe.

vsoch replied:
okay let's briefly mention that the env is cleaned too - even the official [docs](https://sylabs.io/guides/3.1/user-guide/bind_paths_and_mounts.html) don't make that entirely clear.

rhpvorderman replied:
Done! I added a sentence that mentions this happening.

cjllanwarne commented:
@rhpvorderman we can't merge with cancelled tests, but if you only change documentation, our CI should be able to spot that and only run a minimal set (see https://github.com/broadinstitute/cromwell/blob/develop/src/ci/bin/test.inc.sh#L144... we might need to add `CHANGELOG.MD` to that pattern match)

rhpvorderman commented:
@cjllanwarne Thanks for the suggestion. I will do that after I have processed the review comments.

rhpvorderman commented:
@vsoch and @illusional I have processed your comments. I made it more clear why the `--containall` flag is so important and I dropped the list of stuff that Singularity does without the flag. Instead I focused on the way Singularity affects reproducibility and how this can be prevented by the containall flag.

I have also removed any references to a particular version of Singularity, and provided alternatives for module load.

rhpvorderman commented:
@cjllanwarne It still runs all the tests! 

...

And then I realized that altering `src/ci/bin/test.sh` is not included in the regex... Oh well I solved this problem for the next documentation PR then :wink:.

rhpvorderman commented:
@vsoch , @illusional , @TMiguelT any comments that I forgot to process properly? Any additional remarks?

illusional commented:
Hey @rhpvorderman, your changes look fine to me. One thing I noticed is that we've used `${script}` when I believe it should actually be `${docker_script}` to get the container relevant path for the script. Can you confirm in your configurations?

Otherwise I'm happy to click approve from my side (if I'm allowed to do that?)

rhpvorderman commented:
@illusional Thanks! Well, I do not know if you have write access. But I am sure the Cromwell team is happy to know that this PR is approved by other people in the HPC space.

As regards to `docker_script`. It seems that `script` is what gets generated from the WDL task, and it works flawlessly. I don't know what `docker_script` does. It seems to be mentioned only in some corners of the code base and it is not mentioned in the container documentation. My guess would be that this is a now obsolete remnant of some design choices that have been made in the past. `script` works and it is used in all the examples on the container page, so let's be consistent and use that.

illusional commented:
I have write access so technically I can approve, but I’ll leave it to you and @cjllanwarne to decide if my review is enough to merge.

vsoch commented:
@rhpvorderman I still think users should be encouraged to pull a binary first (and then use it) but I've already expressed that.

rhpvorderman commented:
@vsoch This config already does this by using exec. This way the binary image is created in the singularity cache dir.

rhpvorderman commented:
For further clarification. There are two `exec` statements in the config. The `exec` statement **before** the submit command will execute a simple echo command. If singularity notices the image is not there it will pull it to `SINGULARITY_CACHEDIR`. 
Then the job is submitted and the `exec` statement **in** the submit command is executed. If `SINGULARITY_CACHEDIR` is on a shared filesystem the image will already be present. The image will *not* be pulled by the execution node and the job executes right away.

illusional commented:
A follow-up question (prompted by the `quay.io` outage today), does `singularity exec` check the internet to see if it has the most up to date version of the tag? What happens if I gave it a docker with digest instead, would it still poll the internet?

vsoch commented:
I think it has to, in order to see if the image in the cache hash matches the current pull (or other) request, here is an example: https://github.com/hpcng/singularity/blob/2c6cf59870cf0172d61099a3198def8334c94827/internal/pkg/client/library/pull.go#L55. I’m not super familiar with the code, but looks like the hash is retrieved here https://github.com/hpcng/singularity/blob/2c6cf59870cf0172d61099a3198def8334c94827/internal/pkg/client/oci/pull.go#L36 and then needs to get a manifest https://github.com/hpcng/singularity/blob/2c6cf59870cf0172d61099a3198def8334c94827/internal/pkg/build/oci/oci.go#L133 which I’d suspect does just that. https://github.com/containers/image/blob/175bf8b8f9ad897bdc10761e11b466d00f516a63/types/types.go#L238. If a user doesn’t have internet access, or has limited, or there is need to query the registry, might run into trouble. I just tried running an exec to a Docker uri, first of course with Internet to make sure that the images in my cash, and then I disabled my wireless. Without wireless, of course, I couldn’t run anything. This issue could be avoided if the user pulled an image first and then use that image instead of this Docker uri.

vsoch commented:
Here is the terminal output, for posterity:

The first command with internet, confirms we are using a cached image:
```bash
$ singularity exec docker://busybox ls
INFO:    Using cached SIF image
CHANGELOG.md  LICENSE       README.md     dist          paper         qme.egg-info  setup.py
Dockerfile    MANIFEST.in   build         docs          qme           setup.cfg     tests
```
then I took off my wireless :scream: and ran the same - we know the image is in the cache:

```bash
$ singularity exec docker://busybox ls
FATAL:   Unable to handle docker://busybox uri: failed to get checksum for docker://busybox: error pinging docker registry registry-1.docker.io: Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io: no such host
```
```

rhpvorderman commented:
@vsoch Thanks for testing this! This was indeed a major oversight on my behalf. 

I did some further testing:

+ It does not matter if you use a hashed container. Singularity will still look it up on the internet.
+ The singularity cache does not store the file in an easily retrievable way:

```
~/.singularity/cache/oci-tmp/7c7e798af52365c2fa3c1c4606dcf8c1e2d5e502f49f1185d774655648175308$ ls
fastqc_0.11.9--0.sif  fastqc@sha256_319b8d4eca0fc0367d192941f221f7fcd29a6b96996c63cbf8931dbb66e53348.sif
```
You would have to hack with find etc.

Dammit, this means this solution only works for fully connected nodes. And it means an alternate (more robust) solution needs to be hacked together in bash :scream: .

On the other hand, I feel this could be fixed easily by singularity having a `--use-cache-first` flag, so it checks the cache first instead of checking the internet. I will investigate what is possible upstream.

rhpvorderman commented:
It seems that @TMiguelT has already addressed part of [this issue](https://github.com/hpcng/singularity/issues/2597). 

The singularity cache is not meant as a replacement for storing your images. It is merely meant as a way to avoid downloading too much. 

It seems we have to hack something together in bash or contribute something upstream. I think bash hacking can be fairly succesful but it will be very ugly.

The requirements for the cache:
* Make use of singularity's layer cache to avoid too much downloading.
* Should be thread safe. Singularity 3.6.0 commands will use a threadsafe cache. Older version will have to use a universal file lock.
* Will put all `.sif` images in a single shared location.
* Will check if a file exists before pulling. This will mean a cache-first approach. Internet outages should not affect workflows that have already run several times.

rhpvorderman commented:
I made this issue https://github.com/hpcng/singularity/issues/5309 . I will wait for the upstream reaction and decide what to do after that.

rhpvorderman commented:
Since it's still crickets chirping upstream, and quay.io was down again yesterday, I decided to work a bit on this permanent cache thing.

Nobody likes bash scripts. But are python scripts okay? I was wondering what you think about this python script @TMiguelT, @illusional, @vsoch ? The PR where it is created is [here](https://github.com/biowdl/singularity-permanent-cache/pull/1). It is open for feedback.

The python script:
- Pulls images to a single location based on environment or command line flag.
- Uses `singularity pull` as the backend.
- Returns the location of the image to stdout.
- Checks if an image is present in the cache. If so, it returns it and does not use `singularity pull` in that case. It does not require any internet connection in that case.
- Utilizes a filelock to prevent cache corruption. It uses flock to do this.
- Has no dependencies. Only a modern version of python is needed (3.5 or higher).

dctrud commented:
Hello - @vsoch has pointed me over here, and reminded me of https://github.com/hpcng/singularity/issues/5309 which I'll respond to shortly.

It may be useful to note here that in the forthcoming singularity 3.6 the cache functionality is rewritten to drop the additional directory structure and stuff that caused race conditions. We now pull fetch cache entries to a tmp name, and rename - so assuming the filesystem the the cache dir is on supports atomic rename there shouldn't be concurrency issues.

This will not prevent the need to work around it here for older singularity versions, though.

dctrud commented on docs/tutorials/Containers.md:
```diff
@@ -178,27 +190,48 @@ To run Singularity on a job scheduler, the singularity command needs to be passe
-  
+    # Make sure cache dir exists so lock file can be created by flock
+    mkdir -p $CACHE_DIR  
+    LOCK_FILE=$CACHE_DIR/singularity_pull_flock
```
Not sure whether it's relevant (as I don't know if these pulls will run across many nodes), but just wanted to raise that `flock` based approaches for synchronization such as here can come unstuck when locking isn't cluster consistent.

It's not unheard of for people to be running Lustre mounted `localflock` so locking is only node local etc.

vsoch commented:
Thank you @dctrud !

aednichols commented:
This PR is ready to go as far as the approval process does, but we can hold it open if there is continued discussion.
